lua/tint.lua

```diff
@@ -1,121 +1,147 @@
+    bg = hl_def.bg or hl_def.background,
+    sp = hl_def.sp or hl_def.special,
+    blend = hl_def.blend,
+    bold = hl_def.bold,
+    standout = hl_def.standout,
+    underline = hl_def.underline,
+    undercurl = hl_def.undercurl,
+    underdouble = hl_def.underdouble,
+    underdotted = hl_def.underdotted,
+    underdashed = hl_def.underdashed,
+    strikethrough = hl_def.strikethrough,
+    italic = hl_def.italic,
+    reverse = hl_def.reverse,
+    nocombine = hl_def.nocombine,
+    link = hl_def.link,
+    default = hl_def.default,
+    ctermfg = hl_def.ctermfg,
+    ctermbg = hl_def.ctermbg,
+    cterm = hl_def.cterm,
   }
 end
 
-local function ignore_tint(winid)
-  return tint.config.ignorefunc and tint.config.ignorefunc(winid) or false
+--- Get the set of transforms to apply to highlight groups from the colorscheme in question
+---
+---@return table A table of functions to transform the input RGB color values by
+local function get_transforms()
+  if type(__.user_config.transforms) == "string" and __.transforms[__.user_config.transforms] then
+    return __.transforms[__.user_config.transforms]()
+  elseif __.user_config.transforms then
+    return __.user_config.transforms
+  else
+    return __.transforms[tint.transforms.SATURATE_TINT]
+  end
+end
+
+--- Determine if a window should be ignored or not, triggered on `WinLeave`
+---
+---@param winid number Window ID
+---@return boolean Whether or not the window should be ignored for tinting
+local function win_should_ignore_tint(winid)
+  return __.user_config.window_ignore_function and __.user_config.window_ignore_function(winid) or false
 end
 
-local function ignored(k)
-  for _, pat in ipairs(tint.config.ignore) do
-    if string.find(k, pat) then
+--- Determine if a highlight group should be ignored or not
+---
+---@param hl_group_name string The name of the highlight group
+---@return boolean `true` if the group should be ignored, `false` otherwise
+local function hl_group_is_ignored(hl_group_name)
+  for _, pat in ipairs(__.user_config.highlight_ignore_patterns) do
+    if string.find(hl_group_name, pat) then
       return true
     end
   end
 
   return false
 end
 
-local function set_default_ns(k, def)
-  vim.api.nvim_set_hl(__.default_ns, k, def)
+--- Create the "default" (non-tinted) highlight namespace
+---
+---@param hl_group_name string
+---@param hl_def table The highlight definition, see `:h nvim_set_hl`
+local function set_default_ns(hl_group_name, hl_def)
+  vim.api.nvim_set_hl(__.default_ns, hl_group_name, hl_def)
 end
 
-local function set_tint_ns(k, def)
-  if def.fg and not ignored(k) then
-    def.fg = transform_color(get_hex(def.fg), tint.config.amt, tint.config.saturation)
+--- Create the "tinted" highlight namespace
+---
+---@param hl_group_name string
+---@param hl_def table The highlight definition, see `:h nvim_set_hl`
+local function set_tint_ns(hl_group_name, hl_def)
+  local hl_group_info = { hl_group_name = hl_group_name }
```

Comment:
_fyi_: `hl_group_info` defined here - basically just a way for users in their various `transform` functions to do things "dynamically" based of of what color they are transforming. E.g. if they wanted to handle one color differently than another.